src/hooks/wallets/useProvider.ts

```diff
@@ -0,0 +1,34 @@
+import { useMemo } from 'react'
+import type { JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+
+import { useCurrentChain } from '@/hooks/useChains'
+import useWallet from '@/hooks/wallets/useWallet'
+import { createSafeAppsWeb3Provider, createWeb3, createWeb3ReadOnly } from '@/hooks/wallets/web3'
+import { useAppSelector } from '@/store'
+import { selectRpc } from '@/store/settingsSlice'
+
+export const useProvider = (isSafeApp = false): JsonRpcProvider | Web3Provider | undefined => {
+  const chain = useCurrentChain()
+  const wallet = useWallet()
+  const rpcMap = useAppSelector(selectRpc)
+
+  const customRpc = chain ? rpcMap[chain.chainId] : undefined
+
+  return useMemo(() => {
+    if (!chain) {
+      return
+    }
+
+    const isReadOnly = !wallet || wallet.chainId !== chain.chainId
+
+    if (!isReadOnly) {
+      return createWeb3(wallet.provider)
```

Comment:
As most users probably have a wallet connected I would assume that then most of the time the wallet provider will be used.
I would like to understand what was the user story for our custom RPC provider in the Safe Settings as this PR would circumvent this customRpc most of the time now. If privacy reasons / broken rpcs in some wallets were the reason we should consider giving the customRPC higher priority if readonly methods are needed.

Reply:
As far as I understand, with this refactor we now only have one provider at all times be it read-only or not so there shouldn't be any possible desync issues anymore. I agree that for the custom rpc setting it is not clear how it will affect the usage so maybe we should clarify that as you suggested i.e. if the wallet is not connected or if its not connected to the current network.

I think having the custom rpc option is still useful because safe apps can potentially have a lot of rpc requests which would go through the default rpc otherwise.

Reply:
> I think having the custom rpc option is still useful because safe apps can potentially have a lot of rpc requests which would go through the default rpc otherwise.

This would not be the case. If the wallet is connected to the current chain, the wallet RPC would be used.

Reply:
> This would not be the case. If the wallet is connected to the current chain, the wallet RPC would be used.

Is this change intended? Previously, we always used a read-only provider inside the app communicator with its own infura token I assume to keep it separate from rpc requests of the main app.

Reply:
> Is this change intended? Previously, we always used a read-only provider inside the app communicator with its own infura token I assume to keep it separate from rpc requests of the main app.

Yes, otherwise there will different RPCs being used.

Reply:
As per our discussion, a side-effect of this change is that we will most likely see decreased usage of the safe apps infura token which makes sense.